### PR TITLE
C7 0.4.0

### PR DESCRIPTION
**Pull request by paulofaria**
* Rename some APIs to be more inlined with Swift API Design Guidelines
* Remove unneeded typealiases
* Add HTTP errors
* Better naming for HTTP header types
* Minor styling problems
* Make serialize receive a stream to make possible integration with #42 
* Remove router related protocols

### CODE REVIEW

File: Sources/Headers.swift [hunk 1]
```diff
@@ -49,9 +43,9 @@ extension Headers {
         return headers.isEmpty
     }
 
-    public subscript(name: HeaderName) -> HeaderValues {
+    public subscript(name: CaseInsensitiveString) -> Header {
```
Comment (tannernelson) [inline Sources/Headers.swift]: Can we allow them to set a `HeaderConvertible` object.

I want to allow this type of syntax:

```swift
let key: String = "key"
let value: String = "value"
request.headers[key] = value
request.headers.append(value)
```

Comment (tannernelson) [inline Sources/Headers.swift]: Right now I have to do 

```swift
let key: String = "key"
let value: String = "value"
request.headers[key] = Header(value)
request.headers.append(Header(value))
```

Comment (paulofaria) [inline Sources/Headers.swift]: > Can we allow them to set a HeaderConvertible *object*.
Java programmer detected. 🚔 hahaha

I'm working on it.

Comment (tannernelson) [inline Sources/Headers.swift]: I've never done Java in my life officer, I swear. :eyes:

File: Sources/S4.swift [hunk 2]
```diff
@@ -1 +1,9 @@
 @_exported import C7
+
+#if swift(>=3.0)
+#else
+    public typealias ErrorProtocol = ErrorType
+    public typealias RangeReplaceableCollection = RangeReplaceableCollectionType
+    public typealias MutableCollection = MutableCollectionType
+    public typealias Sequence = SequenceType
+#endif
```
Comment (scottbyrns) [inline Sources/S4.swift]: New line

**Comment by scottbyrns** [comment]
LGTM!

**Comment by paulofaria** [comment]
This PR allows this:

```swift
let foo = "foo"
let bar = "bar"
let baz = "baz"
request.headers[foo] += bar
request.headers[key] += baz
```

**Pull request merged**